### PR TITLE
Add JSON schema

### LINKED ISSUE #18: JSON schema
`webman check` does manual validation of recipes, however something like a [JSON schema](https://json-schema.org/) could aid those making recipes on their own, and would also allow `webman check` to simply run a schema lint against recipes.

I would be open to making a PR if you like the proposal.

### PR DESCRIPTION
**Pull request by jolheiser**
Will resolve #18 

This is a draft of the schema, you can add it to your IDE of choice that supports it.
I will be adding some library that can ingest this schema and lint a given recipe via CLI as well.

I had minimized some of the JSON to decrease whitespace noise, but JetBrains had other ideas apparently... :sweat_smile: 

Feel free to give early feedback, particularly for descriptions or if you can think of any odd validation the code does, as I've only looked through the code briefly to look for existing validation rules.

I noticed that you only have mappings for three OS, so let me know if you also want to limit the `enum` for ARCH in the schema. Currently it's just a list I put together from `go tool dist list`, although some obvious ones like `wasm` could probably be safely ignored.

### CODE REVIEW

**Comment by candrewlee14** [comment]
Wow, that was fast, great work! 
I'm not super familiar with the limitations of schema descriptions like this, but I'm wondering if the specific options for the latest_version_strategy could be enforced. For example, if you use the community arch linux strategy, then you should have an arch linux package name, but otherwise, that field shouldn't be there. Can that be enforced through the schema?

If that can't be done unless we change the package format, I'd be willing to weigh the tradeoffs.

**Comment by candrewlee14** [comment]
Also, looks like the end of the README for https://github.com/xeipuuv/gojsonschema addresses YAML schema checking, so that might be a useful library to check out.

**Comment by jolheiser** [comment]
I'll take a look at the library mentioned. 

For the last part, personally my preference is changing the name. I'm ambivalent on `.webman.yml` vs `.pkgrcp.yml` (and iirc we can have it look for both `yaml` vs `yml`)

I know that's some effort on your part with the pkg repo, so I'll let you decide which, if any.

File: schema/schema.json [hunk 1]
```diff
@@ -1,29 +1,27 @@
...
     "filename_format",
     "base_download_url",
-    "latest_strategy"
+    "latest_strategy",
+    "os_map",
+    "arch_map"
   ],
   "additionalProperties": false,
   "properties": {
-    "title": {
```
Comment (jolheiser) [inline schema/schema.json]: This is a required property in a recipe, right?

**Comment by jolheiser** [comment]
So, running the linter gives some errors on the current packages, so I want feedback on whether it needs tweaking or if you agree that the errors are "real"

-----
`pboy.yaml: arch_map.amd64: Invalid type. Expected: string, given: null`
https://github.com/candrewlee14/webman-pkgs/blob/772d53e5e37496d847728a6edd7c58d0bdc0002a/pkgs/pboy.yaml#L31

-----
`lf.yaml: arch_map: Additional property i386 is not allowed`
https://github.com/candrewlee14/webman-pkgs/blob/772d53e5e37496d847728a6edd7c58d0bdc0002a/pkgs/lf.yaml#L31

`lf.yaml: ignore.3.arch: ignore.3.arch must be one of the following: "ppc64", "368", "amd64", "arm", "arm64", "wasm", "mips", "mips64", "mips64le", "mipsle", "ppc64le", "riscv64", "s390x"`
https://github.com/candrewlee14/webman-pkgs/blob/772d53e5e37496d847728a6edd7c58d0bdc0002a/pkgs/lf.yaml#L43
Go only calls it `386`

`lf.yaml: version_format: Invalid type. Expected: string, given: null`
https://github.com/candrewlee14/webman-pkgs/blob/772d53e5e37496d847728a6edd7c58d0bdc0002a/pkgs/lf.yaml#L14
In terms of the schema, things like this would just be left out entirely to use their default afaik

**Comment by candrewlee14** [comment]
I think all of those are mistakes in the recipe, not the schema.
I've got a `schema` branch in the webman-pkgs repo where I've made changes to the recipe files. Try running the linter against those now.

**Comment by jolheiser** [comment]
> I think all of those are mistakes in the recipe, not the schema. I've got a `schema` branch in the webman-pkgs repo where I've made changes to the recipe files. Try running the linter against those now.

Pretty close. 

`386` -> `368` (I see I mistyped in my other comment, my bad!)

Also technically YAML wants to turn it into an int, which is failing. Wrapping it in quotes works, if that's okay.

**Comment by jolheiser** [comment]
I think this is now ready for review, once the integer issue is taken care of with `386` in the recipes.

**Comment by candrewlee14** [comment]
Did you add a linter in `webman check` or is that for another PR?

**Comment by jolheiser** [comment]
> Did you add a linter in `webman check` or is that for another PR?

I replaced the check with the linter. That is, `CheckPkgConfig` now calls `schema.Lint` on a given recipe file.

**Comment by candrewlee14** [comment]
> > Did you add a linter in `webman check` or is that for another PR?
> 
> I replaced the check with the linter. That is, `CheckPkgConfig` now calls `schema.Lint` on a given recipe file.
> 
> As a side note, `webman` could now dogfood CI the pkgs repo and run an action to lint whenever a PR is made to a package. ✨

That's exactly what I was thinking. 
I think first we should submit PRs to https://github.com/schemastore/schemastore/ for a `webman-pkg.json`.
Then we can use https://megalinter.github.io/latest/ in CI.

File: schema/schema.json [hunk 2]
```diff
@@ -1,29 +1,27 @@
...
   "required": [
     "title",
     "tagline",
     "about",
     "filename_format",
     "base_download_url",
-    "latest_strategy"
+    "latest_strategy",
+    "os_map",
+    "arch_map"
```
[the remark this answers is not in the document]
Comment (jolheiser) [reply schema/schema.json:14]: @candrewlee14 I forgot about this, but are they required? If not we may want to remove them from this list.

**Pull request merged**